source/agora/consensus/BlockRewardCalc.d

```diff
@@ -0,0 +1,651 @@
+            if (blocks_accumulated >= height)
+                return CurrentPeriodTup(period, blocks_accumulated == height);
+        }
+        return CurrentPeriodTup(block_rewards.length, false);
+    }
+
+    unittest
+    {
+        immutable block_rewards = [BlockRewardsTup(Amount(100), 7), BlockRewardsTup(Amount(100), 2)];
+        auto block_reward_calc = getBlockRewardCalc(2, 5, block_rewards);
+        assert(block_reward_calc.getCurrentPeriod(Height(0), block_rewards) == CurrentPeriodTup(0, false));
+        assert(block_reward_calc.getCurrentPeriod(Height(7), block_rewards) == CurrentPeriodTup(0, true));
+        assert(block_reward_calc.getCurrentPeriod(Height(8), block_rewards) == CurrentPeriodTup(1, false));
+        assert(block_reward_calc.getCurrentPeriod(Height(9), block_rewards) == CurrentPeriodTup(1, true));
+        assert(block_reward_calc.getCurrentPeriod(Height(10), block_rewards) == CurrentPeriodTup(2, false));
+    }
+
+    /***************************************************************************
+
+    Returns the validator reward for each validator based on how many blocks
+    each validator signed, and also calculates the remainder that cannot be evenly
+    distributed to the validators.
+
+    Params:
+        height = the height of the block that will contain the CoinBase transaction
+        validator_stakes = number of blocks signed by the validator identified by
+                           its public key
+        remainder = the remainder that cannot be evenly distributed to
+                    the validators
+        comply_perc = (#actual signatures / #expected signatures) * 100 on all
+              the blocks for which we want to calculate the block reward
+        indent = number of spaces before log messages
+
+    Returns: the validator reward for each validator based on how many blocks
+    each validator signed
+
+    ***************************************************************************/
+
+    public Amount[Point] getEachValidatorReward (in Height height, ushort[Point] validator_stakes,
+        out Amount remainder, double comply_perc, uint indent = 0) @safe const nothrow
+    {
+        auto indentStr = ' '.repeat(indent).array;
+        log.trace("{}Calculating each validator reward", indentStr);
+        Amount[Point] each_validator_rewards;
+        Amount total_validator_reward = getReducedValidatorReward(height, comply_perc, indent + indentInc);
+        Amount gcd_validator_reward = total_validator_reward;
+
+        ulong total_stakes;
+        foreach (const ref stake; validator_stakes.byValue())
+            total_stakes += stake;
+        assert(total_stakes != 0);
+
+        Amount min_validator_reward = total_validator_reward;
+        min_validator_reward.div(total_stakes);
+
+        Amount actual_total_validator_reward;
+        foreach (const ref pkey_stake; validator_stakes.byKeyValue())
```

Comment:
Why the complicated pattern when you can just do:
```D
foreach (key; const ref value; validator_stakes)
```

?

Reply:
got a nasty DLang bug when tried:
source/agora/consensus/BlockRewardCalc.d(146,9): Error: _aaApply2 is not nothrow
source/agora/consensus/BlockRewardCalc.d(128,26): Error: nothrow function agora.consensus.BlockRewardCalc.BlockRewardCalc.getEachValidatorReward may throw

to reproduce
```D
void f() nothrow
{
    int[string] m;
    foreach(const ref k, v; m)
    {
    }
}
int testmain()
{
    f();
    return 0;
}
```